### ISSUE TITLE
Each upgrade is not compatible with the old bytecode, I'm crazy

### ISSUE THREAD

**Issue by zeropool** [svaarala/duktape#1266]
I hope that every major version of the upgrade can be compatible with the old version of the bytecode, source encryption is very important for me.

**Comment by fatcerberus** [comment]
Bytecode is not part of the versioning guarantees, see `bytecode.rst`.  In particular, this bit:

> Duktape bytecode is version specific and (potentially) config option specific, and may change arbitrarily even in minor releases (but is guaranteed not to change in a patch release, as long as config options are kept the same).

You also probably shouldn't rely on bytecode as an encryption mechanism, as it can be easily decompiled and/or reverse engineered.  Proper encryption of the source code is a much better solution.

**Comment by botvs** [comment]
@fatcerberus My source is running in the hardware, is there other way to replace the bytecode encryption method to protected my code, because the code requires dynamic debugging, so also can not be use source confused encryption, thank you.

**Comment by zeropool** [comment]
@fatcerberus My source is running in the hardware, is there other way to replace the bytecode encryption method to protected my code, because the code requires dynamic debugging, so also can not be use source confused encryption, thank you.

**Comment by harold-b** [comment]
Encryption won't really help if a malicious agent has access to the hardware. At some point the the code has to be decrypted into the the bytecode format which will be in accessible in memory, at which point it could be dumped. Also, if you're decrypting on the target hardware, I assume the key is also present in the hardware, which would also be an exercise in futility. Even if a key is delivered over the air to be decrypted at runtime, if anyone has access to the hardware, the key at some point will also be present in memory. Again, this is pointless anyway since the bytecode will be inevitably loaded into memory eventually.

So in other words, there is no safe way once your code/hardware is the hands of a crafty malicious agent. Your best bet is obfuscation of the identifiers in order to make it harder for the malicious agent to decipher the usage of the byte code. So look into something like UglifyJS2 or Google Closure Compiler.

Even then, this is still [Security through obscurity](https://en.wikipedia.org/wiki/Security_through_obscurity).

**Comment by zeropool** [comment]
@harold-b thank you very much, your reply is very helpful to me !
